src/tls.rs

```diff
@@ -1,6 +1,234 @@
+                keep_alive_enable: kac.enable,
+                keep_alive_idle: kac.idle.as_secs() as i32,
+                keep_alive_interval: kac.interval.as_secs() as i32,
+                keep_alive_count: kac.count as i32,
+            };
+            rcfg.keep_alive_cfg = &mut raw_kac as *mut _;
+        }
+
+        let mut raw_psk: esp_idf_sys::psk_key_hint;
+        if let Some(psk) = cfg.psk_hint_key {
+            raw_psk = esp_idf_sys::psk_key_hint {
+                key: psk.key.as_ptr(),
+                key_size: psk.key.len(),
+                hint: psk.hint.as_ptr(),
+            };
+            rcfg.psk_hint_key = &mut raw_psk as *mut _;
+        }
+
+        rcfg.is_plain_tcp = cfg.is_plain_tcp;
+        rcfg.if_name = std::ptr::null_mut();
+
+        let ret = unsafe {
+            esp_idf_sys::esp_tls_conn_new_sync(
+                host.as_bytes().as_ptr() as *const i8,
+                host.len() as i32,
+                port,
+                &rcfg,
+                tls,
+            )
+        };
+
+        if ret == 1 {
+            Ok(EspTls { raw: tls })
+        } else {
+            unsafe {
+                esp_idf_sys::esp_tls_conn_destroy(tls);
+            }
+
+            Err(Error::ConnectionNotEstablished)
+        }
+    }
+
+    pub fn close(self) {}
+}
+
+impl Read for EspTls {
+    fn read(&mut self, buf: &mut [u8]) -> io::Result<usize> {
+        // cannot call esp_tls_conn_read bc it's inline
+        //let ret = unsafe { esp_idf_sys::esp_tls_conn_read(self.raw, buf.as_mut_ptr(), buf.len()) };
+        let esp_tls = unsafe { std::ptr::read_unaligned(self.raw) };
+        let read_func = esp_tls.read.unwrap();
+        let ret = unsafe { read_func(self.raw, buf.as_mut_ptr() as *mut i8, buf.len()) } as i32;
+        // ESP docs treat 0 as error, but in Rust it's common to return 0 from `Read::read` to indicate eof
+        if ret >= 0 {
+            Ok(ret as usize)
+        } else {
+            Err(io::Error::new(io::ErrorKind::Other, Error::ReadError(ret)))
+        }
+    }
+}
+
+impl Write for EspTls {
+    fn write(&mut self, buf: &[u8]) -> io::Result<usize> {
+        // cannot call esp_tls_conn_write bc it's inline
+        //let ret = unsafe { esp_idf_sys::esp_tls_conn_write(self.raw, buf.as_ptr(), buf.len()) };
+        let esp_tls = unsafe { std::ptr::read_unaligned(self.raw) };
+        let write_func = esp_tls.write.unwrap();
+        let ret = unsafe { write_func(self.raw, buf.as_ptr() as *const i8, buf.len()) } as i32;
+        if ret >= 0 {
+            Ok(ret as usize)
+        } else {
+            Err(io::Error::new(io::ErrorKind::Other, Error::WriteError(ret)))
+        }
+    }
+
+    fn flush(&mut self) -> std::io::Result<()> {
+        Ok(())
+    }
+}
+
+impl Drop for EspTls {
+    fn drop(&mut self) {
+        unsafe {
+            esp_idf_sys::esp_tls_conn_destroy(self.raw);
+        }
+    }
+}
+
+pub struct Config<'a> {
+    pub alpn_protos: Option<&'a [&'a CStr]>,
```

Comment:
Ideally - `&'a str` instead of `&'a CStr`. That might require `alloc` though as the conversion from `&str` to `&CStr` needs `CString`.